gnucash/report/report-system/test/test-report-jqplot.scm

```diff
@@ -0,0 +1,699 @@
+  )
+
+  (test-end "Jqplot - HTML Bar Chart zero-data detection")
+)
+
+;; -----------------------------------------------------------------------
+
+(define (test-line-chart-zero-detection)
+
+  (test-begin "Jqplot - HTML Line Chart zero-data detection")
+
+  (let (
+         (html-line-chart (gnc:make-html-linechart))
+       )
+    ;; no data, no rendering - check that zero values are detected
+    (gnc:html-linechart-set-data! html-line-chart '())
+    (test-assert "HTML Line Chart - zero detection 1" (not (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart))))
+    (gnc:html-linechart-set-data! html-line-chart '(() () ()))
+    (test-assert "HTML Line Chart - zero detection 2" (not (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart))))
+    (gnc:html-linechart-set-data! html-line-chart 0)
+    (test-assert "HTML Line Chart - zero detection 3" (not (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart))))
+    (gnc:html-linechart-set-data! html-line-chart 0.00)
+    (test-assert "HTML Line Chart - zero detection 4" (not (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart))))
+    (gnc:html-linechart-set-data! html-line-chart '((0) (0) (0)))
+    (test-assert "HTML Line Chart - zero detection 5" (not (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart))))
+    (gnc:html-linechart-set-data! html-line-chart '((0.00) (0.00) (0.00)))
+    (test-assert "HTML Line Chart - zero detection 6" (not (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart))))
+    (not (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart)))
+    ;; check that non-zero values are detected
+    (gnc:html-linechart-set-data! html-line-chart 2)
+    (test-assert "HTML Line Chart - zero detection 7" (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart)))
+    (gnc:html-linechart-set-data! html-line-chart 2.00)
+    (test-assert "HTML Line Chart - zero detection 8" (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart)))
+    (gnc:html-linechart-set-data! html-line-chart '((0) (2) (0)))
+    (test-assert "HTML Line Chart - zero detection 9" (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart)))
+    (gnc:html-linechart-set-data! html-line-chart '((0.00) (0.00) (2.00)))
+    (test-assert "HTML Line Chart - zero detection 10" (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart)))
+    (gnc:html-linechart-set-data! html-line-chart '(() () (3)))
+    (test-assert "HTML Line Chart - zero detection 11" (gnc:not-all-zeros (gnc:html-linechart-data html-line-chart)))
+  )
+
+  (test-end "Jqplot - HTML Line Chart zero-data detection")
+)
+
+;; -----------------------------------------------------------------------
+
+(define (test-bar-chart-data)
+
+  (test-begin "Jqplot - HTML Bar Chart Data")
+
+  (let (
+         (html-bar-chart (gnc:make-html-barchart))
+       )
+    ;; prepare test data
+    (gnc:html-barchart-set-data! html-bar-chart '((5.00)))
```

Comment:
You were doing well until now. This tests 4 separate function at one go and on failure requires careful analysis to figure out what's gone wrong.

Instead, populate a chart with a canned set of elements as a fixture and then run each function in its own test to see that it does what it's supposed to,.

Reply:
Agree.
The return values right now are giving good indications in which direction to look if you are familiar with how the table creation works. If not, than your proposal will make it clearer, so let's go for that and do it similar as for the table data tests in PR#391.

This brings me to a another point:
As the table data manipulation is handled similarily in html-barchart.scm, html-linechart.scm and html-table.scm, should that be centralized? E.g. html-barchat.scm and html-linechart.scm are using the functions from html-table.scm?

Is Chris Lam maybe working on this already?
What I am after: The corresponding tests for html-table.scm came along with a bugfix for row extension. I think it will be the same for bar and line chart. If Chris is working on a new concept than I will drop corresponding tests until this is done, and only keep the simple tests as implemented so far.

Reply:
#316 is the current work-in-progress for new chart generation, which eventually aims to replace html-pie/line/bar/scatter. I would definitely NOT write any whole-document string-comparison tests. I'd write a test for each functionality, with good comments what's being tested.

Compare the following test output:
````
[pass] html-chart-add-row! on an empty table
[pass] html-chart-add-row! on a single row-table
[fail] html-chart-add-row! on a large table
 expected: xxx
 obtained: yyy
[pass] html-chart-remove row
````

is 1000% more informative than
````
[fail] html-chart row manipulation tests
````

Reply:
thanks for the quick response and your clear words.
I understand
- having nothing is better than having anything until it is replaced by something proper
- you have not followed my recent changes by which 999% of "my" tests should return expected/obtained values, have a look at #391 - maybe you noticed that I agreed to John's comment
- this PR is a waste of time, it will anyway be replaced by Chris' work, thanks for letting me know

times appear not to be easy for voluntary contributors, actually, seems like you do not want them